imports/lib/schemas/customTypes.ts

```diff
@@ -0,0 +1,68 @@
+import { Meteor } from 'meteor/meteor';
+import { z } from 'zod';
+import { answerify } from '../../model-helpers';
+import { attachCustomJsonSchema } from './generateJsonSchema';
+import { Id } from './regexes';
+
+// Each of these is set to true based on the context in which the schema is
+// being evaluated. They are mutually exclusive (although they can all be false)
+export const IsInsert = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpdate = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpsert = new Meteor.EnvironmentVariable<boolean>();
+
+// In several of these auto-generated fields, we lie to zod about the output
+// type of the field - with type assertions, we say that they are non-nullable.
+// This will be true in the database, even though the transform functions don't
+// always return values (e.g. createdTimestamp will always be a Date, it'll just
+// get set on insertion and not on subsequent updates).
+
+export const stringId = z.string().regex(Id).optional()
+  .transform((v) => v as unknown as string);
+attachCustomJsonSchema(stringId, { bsonType: 'string', pattern: Id.source }, true);
+
+export const createdTimestamp = z.date()
+  .default(() => new Date());
+attachCustomJsonSchema(createdTimestamp, { bsonType: 'date' }, true);
+
+export const updatedTimestamp = z.date().optional()
+  .transform((v) => {
+    if (v) return v;
+    if (IsUpdate.getOrNullIfOutsideFiber()) return new Date();
+    return undefined;
+  });
+attachCustomJsonSchema(updatedTimestamp, { bsonType: 'date' });
+
+export const lastWriteTimestamp = z.date().optional()
+  .transform((v) => {
+    if (v) return v;
+    if (IsInsert.getOrNullIfOutsideFiber() || IsUpdate.getOrNullIfOutsideFiber()) return new Date();
+    return undefined as unknown as Date;
+  });
+attachCustomJsonSchema(lastWriteTimestamp, { bsonType: 'date' });
+
+export const createdUser = z.string().regex(Id).optional()
+  .transform((v) => {
+    if (v) return v;
+    try {
+      if (IsUpsert.get()) return Meteor.userId();
+    } catch (e) {
+      /* ignore */
+    }
+    return undefined as unknown as string;
+  });
+attachCustomJsonSchema(createdUser, { bsonType: 'string', pattern: Id.source }, true);
+
+export const updatedUser = z.string().regex(Id).optional()
+  .transform((v) => {
+    if (v) return v;
+    try {
+      if (IsUpsert.get()) return Meteor.userId();
+    } catch (e) {
+      /* ignore */
+    }
+    return undefined as unknown as string;
```

Comment:
Since `updatedUser` is generally expected to be optional when returned from the DB, do you need the `as unknown as string` here?

Reply:
Yeah, that's right, I got copy-paste happy. (I think/hope I would have eventually caught this in my tests, but will fix)